### PR TITLE
refactor mint doi functionality

### PR DESCRIPTION
**Pull request by kencho51**
# Pull request for issue: #1666
This is a pull request for the following functionalities:

1. Make the Mint DOI button working in the admindataset page by updating the datacite credentials.
2. The Mint DOI button will update the metadata for an existing/minted DOI.
3.  Introduce two new variables `mds_metadata_url` and `mds_doi_url`, it helps to separate the api endpoints, testing or production.
4. Globally replace non exist prefix `10.5072` with the working prefix `10.80027` to the end point `https://doi.test.datacite.org/`, and also update the tests.
5. The prefix `10.5524` is remained for getting the citation from the production url, eg. the citation dropdown box in the public dataset page

## How to test?

0. pre-requisite
Make sure the following variables exist in the gitlab [fork page](https://gitlab.com/groups/gigascience/forks/-/settings/ci_cd), for accessing the datacite sandbox api.

| variable | value | environment |
| -------- | ----- | ------------ | 
| MDS_USERNAME | CNGB.GIGADB | All (default) |
| MDS_PASSWORD | xxxxxxxxxx | All (default) |
| MDS_PREFIX | 10.80027 | All (default) |
| MDS_METADATA_URL | https://mds.test.datacite.org/metadata | All (default) |
| MDS_DOI_URL | https://mds.test.datacite.org/doi | All (default) |

You could log in [datacite sandbox](https://doi.test.datacite.org/) using the above credentials, you can even click `Create DOI` to manually create a DOI at [here](https://doi.test.datacite.org/repositories/cngb.gigadb/dois). 

For the upstream deployment, the values need to be updated in the upstream gitlab variable [page](https://gitlab.com/gigascience/upstream/gigadb-website/-/settings/ci_cd), and the values can be found in the [here](https://gitlab.com/gigascience/cnhk-infra/-/settings/ci_cd) with the names starting `production_mds_`.

1. In dev environment
```
# remove the exiting .secrets file
% cd gigadb-website
# spin up gigadb website, a new .secrets file will be generated
% ./up.sh
# all tests should be passing
% ./tests/unit_runner
% ./tests/functional_runner
% ./tests/unit_function_runner
% ./tests/acceptance_runner
```

Then go to http://gigadb.gigasciencejournal.com/adminDataset/update/id/8 and click the button `Mint DOI`, you would get `This DOI exists in datacite already, no need to mint, but the metadata is updated!`.

If go to http://gigadb.gigasciencejournal.com/adminDataset/update/id/5  or  http://gigadb.gigasciencejournal.com/adminDataset/update/id/2342 and click the button `Mint DOI`, you would get `error minting a DOI: 422, 422`, which means the metadata's format is invalid, and the button is trying to PUT empty data which is not allowed by the endpoint. And the reason why the metadata's format is invalid because those datasets does not have entry in the table `dataset_author` which causing the xml formed does not contain the `<creator></creator>` section.

2. In staging environment
```
# spin up the aws servers
% cd ops/infrastructure/envs/staging
% ../../../scripts/tf_init.sh --project gigascience/forks/project-gigadb-website --env staging
% terraform plan
% terraform apply
% terraform refresh
# Copy ansible files
% ../../../scripts/ansible_init.sh --env staging
# Provision bastion ec2 server and restore RDS with latest database backup
$ env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories bastion_playbook.yml --extra-vars="gigadb_env=staging"
# Provision webapp ec2 server with ansible
$ TF_KEY_NAME=private_ip env OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ansible-playbook -i ../../inventories webapp_playbook.yml --extra-vars="gigadb_env=staging"

# check gitlab CI/CD pipeline, all tests should be passing, including the acceptance tests
```

Then log in the staging gigadb website as an admin, go to `adminDataset/update/id/5` or any dataset, the `Mint DOI` button will either update the metadata or create the DOI depending whether that DOI exists in [datacite sandbox](https://doi.test.datacite.org/repositories/cngb.gigadb/dois).

Describe how the new functionalities can be tested by PR reviewers

## How have functionalities been implemented?

Describe how the new functionalities have been implemented by the
changed code at a high level

The `actionMint()` has been updated with the valid datacite credentials and logic for updating/creating a DOI in the datacite repository. It will first to store the returned http code  `200` or `404` after checking the status the metadata endpoint and also the doi endpoint, it will update the metadata if it is `200`, it will create a new DOI if it is `404`. 
The according on screen message will be then displayed depending on the update or create action.

## Any issues with implementation?

In the datacite setting, each DOI could have [3 states](https://support.datacite.org/docs/doi-states), `Findable`, `Registered`, and `Draft`, and only `Draft` state can be deleted. This makes testing the `Create` action impossible, because we will not create a draft doi in current setting, and the delete action always return `200`  even a draft is deleted multiple times:
```
# Try to create a draft doi
 % curl -X PUT -H "Content-Type:application/xml;charset=UTF-8" -i --user 'CNGB.GIGADB:$PASSWORD' -d @test_100006.xml https://mds.test.datacite.org/metadata/10.80027 
HTTP/2 201 
date: Mon, 04 Mar 2024 03:52:21 GMT
content-type: text/plain; charset=utf-8
location: https://mds.test.datacite.org/metadata/10.80027/6anq-gs19
status: 201 Created
cache-control: max-age=0, private, must-revalidate
vary: Accept-Encoding, Origin
etag: W/"cc1798c43f10cb453aec8ca069b536da"
x-runtime: 0.244795
x-credential-username: CNGB.GIGADB
x-request-id: 8dcf74e1-24e0-4cb5-bbee-bb1dbb262e50
x-powered-by: Phusion Passenger(R) 6.0.19
server: nginx/1.18.0 + Phusion Passenger(R) 6.0.19

OK (10.80027/6ANQ-GS19)
# Delete draft doi
% curl -X DELETE -H "Content-Type:application/plain;charset=UTF-8" -i --user 'CNGB.GIGADB:$PASSWORD' https://mds.test.datacite.org/metadata/10.80027/6ANQ-GS19
HTTP/2 200 
date: Fri, 08 Mar 2024 14:41:55 GMT
content-type: text/plain; charset=utf-8
status: 200 OK
cache-control: max-age=0, private, must-revalidate
vary: Accept-Encoding, Origin
x-credential-username: CNGB.GIGADB
x-request-id: 844dd09f-01f6-474d-90e0-2e3e0587f94e
etag: W/"565339bc4d33d72817b583024112eb7f"
x-runtime: 0.085310
x-powered-by: Phusion Passenger(R) 6.0.19
server: nginx/1.18.0 + Phusion Passenger(R) 6.0.19

OK%                                               
% curl -X DELETE -H "Content-Type:application/plain;charset=UTF-8" -i --user 'CNGB.GIGADB:$PASSWORD' https://mds.test.datacite.org/metadata/10.80027/6ANQ-GS19
HTTP/2 200 
date: Fri, 08 Mar 2024 14:41:58 GMT
content-type: text/plain; charset=utf-8
status: 200 OK
cache-control: max-age=0, private, must-revalidate
vary: Accept-Encoding, Origin
x-credential-username: CNGB.GIGADB
x-request-id: fc64c15b-93b2-4b19-9b92-e8b730d0400d
etag: W/"565339bc4d33d72817b583024112eb7f"
x-runtime: 0.066805
x-powered-by: Phusion Passenger(R) 6.0.19
server: nginx/1.18.0 + Phusion Passenger(R) 6.0.19

OK%                                  
 % curl -X DELETE -H "Content-Type:application/plain;charset=UTF-8" -i --user 'CNGB.GIGADB:$PASSWORD' https://mds.test.datacite.org/metadata/10.80027/6ANQ-GS19
HTTP/2 200 
date: Fri, 08 Mar 2024 14:42:00 GMT
content-type: text/plain; charset=utf-8
status: 200 OK
cache-control: max-age=0, private, must-revalidate
vary: Accept-Encoding, Origin
x-credential-username: CNGB.GIGADB
x-request-id: 6a315720-0ed9-489a-8206-b1646c1822d0
etag: W/"565339bc4d33d72817b583024112eb7f"
x-runtime: 0.076222
x-powered-by: Phusion Passenger(R) 6.0.19
server: nginx/1.18.0 + Phusion Passenger(R) 6.0.19

OK%                                                                             
```

And the draft dataset is still can be found in https://doi.test.datacite.org/repositories/cngb.gigadb/dois?state=draft.

## Any changes to automated tests?

Describe any automated tests that have been developed for the new 
functionalities

Create functional tests `MintDoiCest.php` to test the status return from the datacite sandbox enpoint under different scenarios.

```
% ./up.sh
% docker-compose run --rm test ./vendor/codeception/codeception/codecept run functional MintDoiCest    
[+] Building 0.0s (0/0)                                                                                                                                                    docker:desktop-linux
[+] Building 0.0s (0/0)                                                                                                                                                    docker:desktop-linux
Codeception PHP Testing Framework v3.1.3
Powered by PHPUnit 5.7.27 by Sebastian Bergmann and contributors.
Running with seed: 


Functional Tests (8) -------------------------------------------------------------------------------------------------------------------------------------------------------------------------
✔ MintDoiCest: Get mds config (0.00s)
✔ MintDoiCest: Try get the status return of an existing doi (2.53s)
✔ MintDoiCest: Try get the status return of non existing doi (2.34s)
✔ MintDoiCest: Try create doi when non exist metadata (1.10s)
✔ MintDoiCest: Try create doi with metadata (2.47s)
✔ MintDoiCest: Try create doi with invalid credentials (2.34s)
✔ MintDoiCest: Try update metadata with not recognizde xml (0.89s)
✔ MintDoiCest: Try update metadata with invalid xml (1.38s)
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Time: 47.66 seconds, Memory: 12.00MB

OK (8 tests, 21 assertions)
% docker-compose run --rm codecept run --no-redirect -g mint-doi acceptance -v
[+] Building 0.0s (0/0)                                                                                                                                                                                       docker:desktop-linux
[+] Building 0.0s (0/0)                                                                                                                                                                                       docker:desktop-linux
Codeception PHP Testing Framework v5.0.4 https://helpukrainewin.org
[Groups] mint-doi 

Acceptance Tests (4) --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Modules: \Helper\Acceptance, Asserts, Db, WebDriver
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
✔ form to update dataset details: Update metadata for an existing doi (9.56s)
✔ form to update dataset details: Update metadata for non exist doi (6.91s)
✔ form to update dataset details: Update metadata with invalid metadata format with existing doi (6.83s)
✔ form to update dataset details: Try to create doi with invalid metadata format (6.90s)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Time: 00:44.391, Memory: 12.00 MB

OK (4 tests, 10 assertions)

```

## Any changes to documentation?

None.

## Any technical debt repayment?

Globally replace non exist prefix `10.5072` with the working prefix `10.80027`.

## Any improvements to CI/CD pipeline?

Describe any improvements to the Gitlab pipeline

Update the `unit_func_tests` stage to create artifacts when failure for debugging.

### CODE REVIEW

**Comment by rija** [comment]
Hi @kencho51,

there are a lot of changes in this PR, but the PR description explains nothing about what's going on.

**Comment by kencho51** [comment]
> 
> there are a lot of changes in this PR, but the PR description explains nothing about what's going on.

The description has been updated.

**Review by rija — COMMENTED** [review]
Hi @kencho51, 

**praise**: the happy path for existing DOI works for me locally and on my staging
**praise**: good work on the refactoring and the associated updating and augmenting of the test suite

**issue** (non-blocking): using PHP curl extension is not recommended in modern PHP, it's better to use the already installed GuzzleHTTP extension: much less verbose, tracks features from modern PHP, supports `PSR-*` standards (including client side caching for speeding the connection to web services and reduce traffic) and middle-wares, and is used in other parts of the codebase for making HTTP connections. **Do not make related changes in this PR as it's not blocking and it needs to be generalised and I've created issue #1750 for this**

**issue** (non-blocking): At line 504 of `protected/controllers/AdminDatasetController.php`, `! in_array($dataset->upload_status, $status_array)`, the variable `$status_array` hardcodes a list of statuses instead of using constants defined in `protected/models/Dataset.php`, which has the consequences:
* if we make change to a status mentioned here (which will happen with #1616) in the canonical location, we need to also change it here (duplication and risk of forgetting )
* None of the FUW related statuses are listed here, and yet we shouldn't be able to Mint DOI with any of them. Fortunately, for now those are switched off by the feature flag
**Do not make related changes in this PR as it's not blocking and it needs to be generalised and I've created issue #1751 for this**


**issue**: The users of the "Mint DOI" button won't understand  the HTTP code that's shown when there is an error
**suggestion** : In the final `else` block of the Mint DOI AJAX code in `protected/views/adminDataset/_form.php` , `output.create_md_status` and `output.create_doi_status` could benefit from being hyperlinks (that open in a new tab) to https://support.datacite.org/reference/mds#api-response-codes  (or show the description instead of the code by keeping in a JS variable a hash mapping the possible response codes to their description, or extract the description from the JSON response, any of those approaches should be fine)

**question**: Locally, dataset ID 5 and 2342 don't give the same result, the former shows `error minting a DOI: undefined, undefined`, while the later returns a proper response (422, 422).
The behaviour on my staging was almost similar (the former showed the same issue, but the latter actually minted the DOI)

It feels like an issue to me, but it may not be blocking if we can have an idea of what's happening here? (Also I don't know if it's just me).
Also it seems that in the middle of my testings, our test repository in Datacite was reset so I cannot fully replicate the above. (And it may explain why I could mint that DOI?)


**issue**: the functional test `tryCreateDoiWithMetadata` seems wrongly named, it's only testing posting an existing DOI (100006) isn't it?  Otherwise we will get confused later when we come back to it. 
**suggestion**: better call it `tryCreateExistingDOIWithMetadata`
And it's probably best adding a TODO for the missing scenario `tryCreateNonExistingDoiWithMetadata` (I created issue #1752 to create a fake MDS API, but if Datacite API can regularly reset test repository - see above -,  it's slightly less critical now because we can just use random DOI)

**praise**: I was able to verify manually the minting of a DOI that doesn't exist yet work, by setting an existing dataset to 'Private' and change (and save) the DOI to something new before clicking the "Mint DOI" button (it showed `new DOI successfully minted` message and the DataCite dashboard was showing it)
**praise**: all tests passing locally
**praise**: all tests passing in CI

**Comment by kencho51** [comment]
> **issue** (non-blocking): using PHP curl extension is not recommended in modern PHP, it's better to use the already installed

I have thought of that too, thanks for creating ticket #1750 for it.

>**issue**: The users of the "Mint DOI" button won't understand the HTTP code that's shown when there is an error

I have refactored the AJAX code to make it more robust in recognizing different HTTP code returned and display the either the response or a link to the datacite api response codes page. And accordingly update the acceptance tests.

>**question**: Locally, dataset ID 5 and 2342 don't give the same result, the former shows error minting a DOI: undefined, undefined, while the later returns a proper response (422, 422).

Please refer to the updated acceptance test with tag `mint-doi`.

>**issue**: the functional test tryCreateDoiWithMetadata seems wrongly named, it's only testing posting an existing DOI 

The name is updated, and an empty function `tryCreateExistingDOIWithMetadata` with TODO tag is added.

**Review by rija — APPROVED** [review]
Hi @kencho51,

**praise**: the additional test scenarios are great addition for covering more edge cases
**praise**: the error message below the button are friendlier an clearer
**praise**: great to see clear test intents in the functional test suite and the TODO to remember there is a missing case


Based on the above I'm happy to approve this PR

File: protected/views/adminDataset/_form.php [hunk 1]
```diff
@@ -349,23 +349,27 @@
...
+                                            'success' => new CJavaScriptExpression('function(output) {
+                                                console.log(output);
+                                                if (output.check_metadata_status == 200 && output.check_doi_status == 200 && output.update_md_status == 201) {
+                                                    $("#minting").addClass("alert alert-info").html("This DOI exists in datacite already, no need to mint, but the metadata is updated!");
+                                                } else if (output.check_metadata_status == 200 && output.check_doi_status == 200 && output.update_md_status == 422) {
+                                                    $("#minting").addClass("alert alert-info").html("This DOI exists in datacite, but failed to update metadata because of: " + output.update_md_response);
+                                                } else if (output.create_md_status == 201 && output.create_doi_status == 201) {
+                                                    $("#minting").addClass("alert alert-success").html("new DOI successfully minted");
```
Comment (pli888) [inline protected/views/adminDataset/_form.php]: It would be good to capitalize the first letter of `new` as this message is shown to user.

**Review by pli888 — APPROVED** [review]
**praise**: Functionality is working as described and code looks good to me, albeit with @rija's comments regarding use of PHP curl and dataset statuses
**chore**: Update `docs/variables.md` with MDS_METADATA_URL and MDS_DOI_URL in SUB-GROUP: Forks section
**question**: I was wondering how this Mint DOI button functionality would work in gigadb.org? For the Upstream gigadb-website project, we would need MDS_* variables for dev, staging and live environments? The live environment would have the API endpoint for the production datacite service, whilst the other upstream environment will use the sandbox datacite API URL? Maybe this is the subject of another PR?
**chore**: Need to rebase with upstream develop branch to resolve conflict in CHANGELOG.md.

**Comment by kencho51** [comment]
> **question**: I was wondering how this Mint DOI button functionality would work in gigadb.org?

As mentioned in the description:
For the upstream deployment, the values need to be updated in the upstream gitlab variable [page](https://gitlab.com/gigascience/upstream/gigadb-website/-/settings/ci_cd), and the values can be found in the [here](https://gitlab.com/gigascience/cnhk-infra/-/settings/ci_cd) with the names starting production_mds_.

May be in the upstream variable page, these mds_* variables need to be created with either `staging` or `live` environment. The staging one using the datacite sandbox variable set, the live one using the real datacite variable found in the [here](https://gitlab.com/gigascience/cnhk-infra/-/settings/ci_cd) with the names starting production_mds_

**Comment by rija** [comment]
> > **question**: I was wondering how this Mint DOI button functionality would work in gigadb.org?
> 
> As mentioned in the description: For the upstream deployment, the values need to be updated in the upstream gitlab variable [page](https://gitlab.com/gigascience/upstream/gigadb-website/-/settings/ci_cd), and the values can be found in the [here](https://gitlab.com/gigascience/cnhk-infra/-/settings/ci_cd) with the names starting production_mds_.
> 
> May be in the upstream variable page, these mds_* variables need to be created with either `staging` or `live` environment. The staging one using the datacite sandbox variable set, the live one using the real datacite variable found in the [here](https://gitlab.com/gigascience/cnhk-infra/-/settings/ci_cd) with the names starting production_mds_

That's the idea indeed, only the environment called `live` will use the real MDS credentials. All other environments will use the test credentials.
